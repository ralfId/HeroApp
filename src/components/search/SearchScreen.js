import React, { useMemo } from 'react'
import { useForms } from '../../customhooks/useForm';
import { HeroCard } from '../heroes/HeroCard'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {
    const location = useLocation();
    const { q = '' } = queryString.parse(location.search);

    const [formValues, handleOnChange] = useForms({ searchText: q })
    const { searchText } = formValues;

    const heroFiltered = useMemo(() => getHeroesByName(q), [q]);


    const handleSubmit = (e) => {
        e.preventDefault();

        history.push(`?q=${searchText}`);
    }




    return (
        <div className="row">
            <div className="col-4">
                <h3>Search Form</h3>

                <form onSubmit={handleSubmit}>
                    <input type="text" placeholder="find a hero" className="form-control" name="searchText" onChange={handleOnChange} />
                    <button className="btn btn-primary  mt-2">search</button>
                </form>
            </div>
            <div className="col-8">
                <h3 className="px-auto">Result</h3>
                <hr />

                {
                    (q === '')
                    &&
                    <div className="alert alert-info">
                        Search a hero
                    </div>
                }
                {
                    (q !== '' && heroFiltered.length === 0)
                    &&
                    <div className="alert alert-danger">
                        There is no a hero called: "{ q }"
                    </div>
                }
                {
                    heroFiltered.map(h => (<HeroCard key={h.id} {...h} />))
                }
            </div>
        </div>
    )
}
