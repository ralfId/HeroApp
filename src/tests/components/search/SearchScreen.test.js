import { mount } from 'enzyme';
import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';


describe('pruebas en SearchScreen.js', () => {

    test('debe de mostrarse correctamente con los valores por defectp', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert-info').exists()).toBe(true)
    });

    test('debe mostrar a batman y el input con el valor del  queryString', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        // expect(wrapper.find('input').prop('value')).toBe('batman')
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('input').exists()).toBe(true)
        expect(wrapper.find('HeroCard').exists()).toBe(true)
    });

    test('debe de mostrar un error si no encuentra el hero', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batmanrrrrr']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-danger').exists()).toBe(true)
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`There is no a hero called: "batmanrrrrr"`)
    });

    test('debe de llamar el push del history', () => {

        const historyMock = {
            push: jest.fn()
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={() => <SearchScreen history={historyMock} />} />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        })

        wrapper.find('form').prop('onSubmit')({ preventDefault(){} })
        
        expect(historyMock.push).toHaveBeenCalledTimes(1)
        expect(historyMock.push).toHaveBeenCalledWith('?q=batman')

    });
});