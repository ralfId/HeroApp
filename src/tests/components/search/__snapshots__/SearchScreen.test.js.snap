// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en SearchScreen.js debe de mostrar un error si no encuentra el hero 1`] = `
<div
  className="row"
>
  <div
    className="col-4"
  >
    <h3>
      Search Form
    </h3>
    <form
      onSubmit={[Function]}
    >
      <input
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="find a hero"
        type="text"
      />
      <button
        className="btn btn-primary  mt-2"
      >
        search
      </button>
    </form>
  </div>
  <div
    className="col-8"
  >
    <h3
      className="px-auto"
    >
      Result
    </h3>
    <hr />
    <div
      className="alert alert-danger"
    >
      There is no a hero called: "
      batmanrrrrr
      "
    </div>
  </div>
</div>
`;

exports[`pruebas en SearchScreen.js debe de mostrarse correctamente con los valores por defectp 1`] = `
<div
  className="row"
>
  <div
    className="col-4"
  >
    <h3>
      Search Form
    </h3>
    <form
      onSubmit={[Function]}
    >
      <input
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="find a hero"
        type="text"
      />
      <button
        className="btn btn-primary  mt-2"
      >
        search
      </button>
    </form>
  </div>
  <div
    className="col-8"
  >
    <h3
      className="px-auto"
    >
      Result
    </h3>
    <hr />
    <div
      className="alert alert-info"
    >
      Search a hero
    </div>
  </div>
</div>
`;

exports[`pruebas en SearchScreen.js debe mostrar a batman y el input con el valor del  queryString 1`] = `
<div
  className="row"
>
  <div
    className="col-4"
  >
    <h3>
      Search Form
    </h3>
    <form
      onSubmit={[Function]}
    >
      <input
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="find a hero"
        type="text"
      />
      <button
        className="btn btn-primary  mt-2"
      >
        search
      </button>
    </form>
  </div>
  <div
    className="col-8"
  >
    <h3
      className="px-auto"
    >
      Result
    </h3>
    <hr />
    <div
      className="col animate__fadeIn"
    >
      <div
        className="card h-100"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img img-fluid"
              src="../assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8 "
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                Detective Comics #27
              </p>
              <a
                className="btn btn-outline-primary"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
